### PR TITLE
fix: make isFocusedOrInFocused check shadow roots

### LINKED ISSUE #26198: The type command should not click on focused elements inside of shadow roots.
### Current behavior

When you attempt to `.type()` inside an element that is inside a shadow root, it will *always* [perform the `.click()`](https://docs.cypress.io/api/commands/type#When-element-is-not-in-focus) even if the element has focus.

This is even an issue if you have `includeShadowDom` set to `true`.

### Desired behavior

Either check inside of shadow roots for the focused element, or allow the `.click()` behavior to be turned off using an option inside the `.type()` command.

### Test code to reproduce

https://github.com/kgroat/cypress-type-repro

### Cypress Version

12.8.1

### Node version

16.16.0

### Operating System

MacOS 13.2.1

### Debug Logs

_No response_

### Other

_No response_

### PR DESCRIPTION
**Pull request by kgroat**
- Closes https://github.com/cypress-io/cypress/issues/26198 

### Additional details
Currently, checking for focus doesn't look inside of shadow roots.  This causes issues such as the `type` command always [firing the `click`](https://docs.cypress.io/api/commands/type#When-element-is-not-in-focus) even if the element has focus.

This PR changes the behavior of `isFocusedOrInFocused` to check if the element passed in is an `activeElement` inside of a shadow root.

### Steps to test
Clone https://github.com/kgroat/cypress-type-repro and run the component test.  Before these changes, it fails, but after it passes.

### How has the user experience changed?
N/A

### PR Tasks

- [x] Have tests been added/updated?
- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

### CODE REVIEW

**Comment by kgroat** [comment]
@lmiller1990 I see that there are several errors from the checks running.  Are they blocking review?  If so, I don't see how they're caused by my code change.  What do I need to do to resolve them?

File: packages/driver/src/dom/elements/complexElements.ts [hunk 1]
```diff
@@ -161,6 +161,10 @@ export const isFocusedOrInFocused = (el: HTMLElement) => {
 
   debug('elToCheckCurrentlyFocused', elToCheckCurrentlyFocused)
 
+  while (elToCheckCurrentlyFocused !== activeElement && activeElement?.shadowRoot?.activeElement) {
```
Comment (emilyrohrbough) [inline packages/driver/src/dom/elements/complexElements.ts]: Is it possible to check to check if the element is within the shadow root before pulling out `doc` on line 143 like we do with `isFocused()` here https://github.com/cypress-io/cypress/blob/develop/packages/driver/src/dom/elements/complexElements.ts#L112?
```js
    let doc
    if (isWithinShadowRoot(el)) {
      doc = el.getRootNode()
    } else {
      doc = $document.getDocumentFromElement(el)
    }
```

File: packages/driver/cypress/e2e/dom/elements.cy.ts [hunk 2]
```diff
@@ -321,4 +321,14 @@ describe('src/dom/elements', () => {
       expect(elementFromPoint(doc, 1, 2)).to.equal(shadowHostNode)
     })
   })
+
+  context('.isFocusedOrInFocused', () => {
+    it('should traverse shadow roots when determining if an element has focus', () => {
+      cy.visit('/fixtures/shadow-dom-focus.html')
```
Comment (emilyrohrbough) [inline packages/driver/cypress/e2e/dom/elements.cy.ts]: Is it possible to use [`/fixtures/shadow-dom-type.html`](https://github.com/cypress-io/cypress/blob/develop/packages/driver/cypress/fixtures/shadow-dom-type.html) to verify this behavior? Would limit the need for an additional fixture.

File: cli/CHANGELOG.md [hunk 3]
```diff
@@ -7,6 +7,7 @@ _Released 03/28/2023 (PENDING)_
 
  - Fixed a compatibility issue so that component test projects can use [Vite](https://vitejs.dev/) version 4.2.0 and greater. Fixes [#26138](https://github.com/cypress-io/cypress/issues/26138).
  - Changed the way that Git hashes are loaded so that non-relevant runs are excluded from the Debug page. Fixes [#26058](https://github.com/cypress-io/cypress/issues/26058).
+ - When checking if an element is focused for `.type()` to determine whether to click, shadow roots are now traversed. Fixes [#26198](https://github.com/cypress-io/cypress/issues/26198).
```
Comment (emilyrohrbough) [inline cli/CHANGELOG.md]:
```suggestion
 - Corrected the [`.type()`](https://docs.cypress.io/api/commands/type) command to account for shadow root elements when determining whether or not focus needs to be simulated  before typing. Fixes [#26198](https://github.com/cypress-io/cypress/issues/26198).
```

File: packages/driver/cypress/e2e/dom/elements.cy.ts (hunk 2 above)
Comment (kgroat) [reply packages/driver/cypress/e2e/dom/elements.cy.ts]: Updated.

File: packages/driver/src/dom/elements/complexElements.ts (hunk 1 above)
Comment (kgroat) [reply packages/driver/src/dom/elements/complexElements.ts]: Good call; updated.

**Comment by kgroat** [comment]
Oh, I clicked the "re-request review" button and it removed @lmiller1990 as a reviewer -- not sure why that happened; I'll make sure not to click that button again 😅

File: packages/driver/src/dom/elements/complexElements.ts [hunk 4]
```diff
@@ -139,14 +139,21 @@ export const isFocusable = ($el: JQuery<HTMLElement>) => {
 
 export const isFocusedOrInFocused = (el: HTMLElement) => {
   debug('isFocusedOrInFocus', el)
-
-  const doc = $document.getDocumentFromElement(el)
+  let doc = $document.getDocumentFromElement(el)
 
   if (!doc.hasFocus()) {
```
Comment (emilyrohrbough) [inline packages/driver/src/dom/elements/complexElements.ts:144]: Would this return pre-maturely if the `el.getRootNode()` didn't have focus?

Comment (kgroat) [reply packages/driver/src/dom/elements/complexElements.ts:144]: I don't think so -- my understanding is that an element inside of a shadow root can only have focus if its shadow root also has focus, all the way up the chain to the document.  So if the document has focus, then some element inside of it (maybe in a shadow root) has focus.

**Review by emilyrohrbough — APPROVED** [review]

**Review by chrisbreiding — APPROVED** [review]

File: cli/CHANGELOG.md [hunk 5]
```diff
@@ -7,6 +7,7 @@ _Released 03/28/2023 (PENDING)_
 
  - Fixed a compatibility issue so that component test projects can use [Vite](https://vitejs.dev/) version 4.2.0 and greater. Fixes [#26138](https://github.com/cypress-io/cypress/issues/26138).
  - Changed the way that Git hashes are loaded so that non-relevant runs are excluded from the Debug page. Fixes [#26058](https://github.com/cypress-io/cypress/issues/26058).
+ - Corrected the [`.type()`](https://docs.cypress.io/api/commands/type) command to account for shadow root elements when determining whether or not focus needs to be simulated  before typing. Fixes [#26198](https://github.com/cypress-io/cypress/issues/26198).
```
Comment (emilyrohrbough) [inline cli/CHANGELOG.md]:
```suggestion
 - Corrected the [`.type()`](https://docs.cypress.io/api/commands/type) command to account for shadow root elements when determining whether or not focus needs to be simulated before typing. Fixes [#26198](https://github.com/cypress-io/cypress/issues/26198).
```

**Comment by kgroat** [comment]
Is there any chance of this getting merged before the release planned tomorrow?

**Comment by lmiller1990** [comment]
Two approvals, CI green, let's ship it! Thanks for the contribution 🎉

**Pull request merged**